app/models/search/index_sweeper.rb

```diff
@@ -35,74 +37,74 @@ def process_batch
     end
   end
 
-  def process_document(item)
-    document = item[item.keys.first] # update/index/delete
-    document_stamp = { document["_id"].to_s => document["error"] }
-
-    first_store = get_store_as_json "first"
-    second_store = get_store_as_json "second"
-    permanent_store = get_store_as_json "permanent"
-
-    if !document["error"]
-      [first_store, second_store, permanent_store].each do |raw_store|
-         raw_store.select  { |doc| doc.keys.first == document["_id"] }.each do |doc|
-           raw_store.delete(doc)
-         end
-       end
+  # Returns a list of all permanent failures associated with the given indexer.
+  # Used for testing purposes. (Can be used for diagnostic purposes, as well.)
+  def self.permanent_failures(indexer)
+    failures = []
 
-      set_store "first", first_store
-      set_store "second", second_store
-      set_store "permanent", permanent_store
-
-      @success_ids << document["_id"]
-
-      return
+    REDIS.hgetall("#{indexer}:failures").each_pair do |id, value|
+      JSON.parse(value).each do |info|
+        if info["count"] >= 3
+          failures << { id.to_s => info["error"] }
+        end
+      end
     end
 
-    return if permanent_store.include?(document_stamp)
-
-    if first_store.include?(document_stamp)
-      second_store << document_stamp
-      first_store.delete(document_stamp)
-      set_store "second", second_store
-      set_store "first", first_store
-      @rerun_ids << document["_id"]
-    elsif second_store.include?(document_stamp)
-      permanent_store << document_stamp
-      second_store.delete(document_stamp)
-      set_store "permanent", permanent_store
-      set_store "second", second_store
-    else
-      first_store << document_stamp
-      set_store "first", first_store
-      @rerun_ids << document["_id"]
-    end
+    failures
   end
 
   private
 
-  def ensure_failure_stores
-    ["first",
-     "second",
-     "permanent"].each do |store_name|
-      unless REDIS.get("#{@indexer}:#{store_name}_failure_store")
-        set_store store_name, []
-      end
-    end
+  # Calculate which IDs were included in this batch.
+  def batch_ids
+    @batch_ids ||= @batch["items"].map { |item| item.values.first["_id"].to_s }
   end
 
-  def all_stores_empty?
-    get_store_as_json("first").empty? &&
-      get_store_as_json("second").empty? &&
-      get_store_as_json("permanent").empty?
+  # Load information about previous errors for all the items in this batch.
+  def load_errors
+    errors = REDIS.hmget("#{@indexer}:failures", batch_ids).
```

Comment:
If you use `mapped_hmget` and `mapped_hmset` you won't have to do the array-hash conversions yourself.

Reply:
Thanks for the suggestion! I changed the two calls in IndexSweeper.